app/models/tag.rb

```diff
@@ -219,29 +188,85 @@ def lists_by_entity_count(page = 1)
+               taggings.created_at as tagging_created_at,
+               COALESCE(entity.updated_at, ls_list.updated_at, relationship.updated_at) AS event_timestamp,
+               COALESCE(entity.last_user_id, ls_list.last_user_id, relationship.last_user_id) AS editor_id,
+               'tagable_updated' as event
+	FROM taggings
+	LEFT JOIN entity ON taggings.tagable_id = entity.id AND taggings.tagable_class = 'Entity'
+ 	LEFT JOIN ls_list ON taggings.tagable_id = ls_list.id AND taggings.tagable_class = 'List'
+	LEFT JOIN relationship ON taggings.tagable_id = relationship.id AND taggings.tagable_class = 'Relationship'
+	WHERE taggings.tag_id = #{id}
+            # adding a tag triggers a callback that also updates the tagable. This clause excludes those updates
+	    AND abs(TIMESTAMPDIFF(SECOND, taggings.created_at, COALESCE(entity.updated_at, ls_list.updated_at, relationship.updated_at))) > 100
+      )
+        UNION
+      (
+        SELECT taggings.id as tagging_id,
+	       taggings.tagable_id,
+	       taggings.tagable_class,
+	       taggings.created_at AS tagging_created_at,
+	       taggings.created_at AS event_timestamp,
+               taggings.last_user_id AS editor_id,
+	       'tag_added' AS event
+        FROM taggings
+	WHERE taggings.tag_id = #{id}
+      )
+        ORDER BY event_timestamp DESC
+        LIMIT #{PER_PAGE}
+        OFFSET #{(page.to_i - 1) * PER_PAGE}
+    SQL
+
+    # NOTE: our version of Mysql2::Result is missing the very convenient method: #to_hash ...why?
+    # we should be able to do ActiveRecord::Base.connection.execute(sql).to_hash instead of
+    # populating the array ourselves...
+    result = []
+    ActiveRecord::Base.connection.execute(sql).each(:as => :hash) { |h| result << h }
+    result
   end
 
-  # [ Hash ] -> Hash
-  #  example: output:
-  # {
-  #  "Relationship" => { 12 => <Relationship>, 22 => <Relationship> }
-  #  "Entity" => { 123 => <Entity>}
-  #  "List" => { 987 => <List> }
+  # type TagablesByClassAndId = {
+  #   "Relationship" => { [id: Integer] => Relationship }
+  #   "Entity"       => { [id: Integer] => Entity }
+  #   "List"         => { [id: Integer] => List }
   # }
-  #
-  def active_record_lookup_for_recent_edits(edits)
-    edits
+  # [EditsIdHash] -> TagablesByClassAndId
+  def tagables_by_class_and_id_for(edits_id_hash)
+    base = { "Relationship" => {}, "Entity" => {}, "List" => {} }
+    edits_id_hash
       .group_by { |h| h['tagable_class'] }
       .transform_values { |tagable_array| tagable_array.map { |h| h['tagable_id'] }.uniq }
       .to_a
-      .reduce("Relationship" => {}, "Entity" => {}, "List" => {}) do |acc, (klass, tagable_ids)|
+      .reduce(base) do |acc, (klass, tagable_ids)|
         klass.constantize.find(tagable_ids).each { |tagable| acc[klass].store(tagable.id, tagable) }
         acc
       end
   end
+
+  # type EditorsById = { [id: Integer] => User }
+  # [EditsIdHash] => EditorsById
+  def editors_by_id(id_hashes)
```

Comment:
cool! I'm curious why you didn't choose to add User (or SfGuardUser) to an adapted version of the the lookup table generator and adapt the function ``` tagables_by_class_and_id_for ``` (which was previously named active_record_lookup_for_recent_edits) ?

Reply:
interesting question. it hadn't occured to me to do that.

i guess i did it here because 

(1) the type of data being retrieved is substantially different - this function returns grouped users whereas the other one returns grouped `Tagables`
(2) the logic for finding, transforming, and hash-table-storing `SfGuardUsers` is substantially different than that used for finding/tranforming/storing `Users` that it would require new branching logic in any function that wanted to do both. specifically, you'd probably need some sort of branch on line 257, which would either (a) make that function get more unwiedly or (b) require splitting out two helper functions for that step -- in which case, why not just have two helpers to start with?

in any case, since this "return ids from SQL, then make a batch lookup table, then retrieve the records using the ids" seems to becoming a bit of a LS design pattern, i think it might be worth thinking about how to generalize it to work for a bunch of assorted ActiveRecord models. willl think about a more generalized implementation next time i have the need to use the lookup table pattern.

(for the time being it was honestly just the fact that what you'd written was so specific to tagables as to seem to require enough work to make it work for something else that nudged me to write a new function to have space to write something that i knew would work for data with different shapes. next step in generalizing would be to try to find a common function that expresses what both of these express in more general terms. would be interested to think about that next time it comes up.)